scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/fgboost/VflGBoostAggregator.scala

```diff
@@ -12,24 +31,61 @@ import scala.collection.JavaConversions._
 import scala.collection.JavaConverters._
 import scala.collection.mutable.ArrayBuffer
 
-class VflGBoostAggregator extends Aggregator[Table] {
+class VflGBoostAggregator(validationMethods: Array[ValidationMethod[Float]] = null)
+  extends Aggregator {
 
   val logger = LogManager.getLogger(this.getClass)
 
   var obj: TreeObjective = new RMSEObjective
   var nLabel = 1
 
-  val trainMap = Map[String, Table]()
-  protected val bestSplit = new HashMap[String, DataSplit]
-  protected val leafMap = new HashMap[String, TreeLeaves]
-  protected val splitMap = new HashMap[String, DataSplit]
-  protected val gradMap = new HashMap[String, List[Float]]
-  val _evalMap = new HashMap[String, java.util.List[BoostEval]]
-  val _predMap = new HashMap[String, java.util.List[BoostEval]]
   val serverTreeLeaves = new ArrayBuffer[Map[Int, Float]]()
   var validationSize = -1
   var basePrediction: Array[Float] = null
+  val bestSplit = new java.util.HashMap[String, DataSplit]()
+  var validationResult = Array[ValidationResult]()
+
+  override def initStorage(): Unit = {
+    aggregateTypeMap.put(FLPhase.LABEL, new StorageHolder(FLDataType.TABLE))
+    aggregateTypeMap.put(FLPhase.PREDICT, new StorageHolder(FLDataType.TABLE))
+    aggregateTypeMap.put(FLPhase.SPLIT, new StorageHolder(FLDataType.SPLIT))
+  }
 
+  override def aggregate(flPhase: FLPhase): Unit = {
+    // TODO aggregate split
+    flPhase match {
+      case FLPhase.LABEL => initGradient()
+      case FLPhase.SPLIT => aggregateSplit()
+      case FLPhase.LEAF => aggregateTree()
+      case FLPhase.EVAL => aggEvaluate()
+      case FLPhase.PREDICT => aggPredict()
+      case _ => throw new NotImplementedError()
+    }
+  }
+
+  def getBestSplit() = {
+    val splitClientData = aggregateTypeMap.get(FLPhase.SPLIT).getSplitStorage().clientData
```

Comment:
this is too complex

Reply:
This is wrapped by some high level method like https://github.com/Litchilitchy/BigDL/blob/03084e2eae5ab6de83fe41fa16d2c458420e8dc1/scala/ppml/src/main/scala/com/intel/analytics/bigdl/ppml/fgboost/FGBoostAggregator.scala#L49